sdk/rust/oak/src/grpc/mod.rs

```diff
@@ -162,6 +162,94 @@ pub trait OakNode {
     fn invoke(&mut self, method: &str, req: &[u8], writer: ChannelResponseWriter);
 }
 
+/// Encapsulate a protocol buffer message in a GrpcRequest wrapper using the
+/// given method name.
+pub fn encap_request<T: protobuf::Message>(req: T, method_name: &str) -> Option<Vec<u8>> {
+    // Put the request in a GrpcRequest wrapper and serialize it.
+    let mut grpc_req = GrpcRequest::new();
+    grpc_req.set_method_name(method_name.to_string());
+    let mut any = protobuf::well_known_types::Any::new();
+    if let Err(e) = req.write_to_writer(&mut any.value) {
+        warn!("failed to serialize gRPC request: {}", e);
+        return None;
+    };
+    grpc_req.set_req_msg(any);
+    grpc_req.set_last(true);
+    let mut buf = Vec::new();
+    grpc_req
+        .write_to_writer(&mut buf)
+        .expect("failed to serialize GrpcRequest message");
+    Some(buf)
+}
+
+/// Helper to inject a (single) gRPC request message via a notification channel,
+/// in the same manner as the gRPC pseudo-Node does, and collect a (single)
+/// response.
+pub fn invoke_grpc_method<R, Q>(
+    method_name: &str,
+    req: R,
+    invocation_channel: &crate::io::Sender<Invocation>,
+) -> Result<Q>
+where
+    R: protobuf::Message,
+    Q: protobuf::Message,
+{
+    // Put the request in a GrpcRequest wrapper and serialize into a message.
+    let req_data = encap_request(req, method_name).expect("failed to serialize GrpcRequest");
+
+    // Create a new channel to hold the request message.
+    let (req_write_half, req_read_half) =
+        crate::channel_create().expect("failed to create channel");
+    crate::channel_write(req_write_half, &req_data, &[]).expect("failed to write to channel");
+    let _ = crate::channel_close(req_write_half.handle);
+
+    // Create a new channel for responses to arrive on and also attach that to the message.
+    let (rsp_write_half, rsp_read_half) =
+        crate::channel_create().expect("failed to create channel");
+
+    let invocation = Invocation {
+        request_receiver: crate::io::Receiver::<GrpcRequest>::new(req_read_half),
+        response_sender: crate::io::Sender::<GrpcResponse>::new(rsp_write_half),
+    };
+
+    invocation_channel
+        .send(&invocation)
+        .expect("failed to write invocation to channel");
+    let _ = crate::channel_close(req_read_half.handle);
+    let _ = crate::channel_close(rsp_write_half.handle);
+
+    // Read a single serialized, encapsulated response.
+    if let Err(status) = crate::wait_on_channels(&[rsp_read_half]) {
+        error!("failed to wait for response: {}", status.value());
+        return Err(build_status(
+            Code::INTERNAL,
+            &format!("failed to wait for gRPC response: {}", status.value()),
+        ));
+    }
+    let mut buf = Vec::with_capacity(1024);
+    let mut handles = Vec::new();
+    let result = crate::channel_read(rsp_read_half, &mut buf, &mut handles);
```

Comment:
This should also use the `Receiver` version. I expect raw `channel_read` to be used less and less going forward (and in fact perhaps we should consider moving it to a `sys` module or similar?).

Reply:
Done.